RPC: listunspent, add "include immature coinbase" flag

furszy opened the pull request:
Simple PR; adds a "include_immature_coinbase" flag to `listunspent` to include the immature coinbase UTXOs on the response.  Requested by #25728.

furszy committed:
RPC: listunspent, add "include immature coinbase" flag

so we can return the immature coinbase UTXOs as well.

w0xlt reviewed (approved):
Tested ACK https://github.com/bitcoin/bitcoin/pull/25730/commits/8621fbf672157df5475fc5cbaa6524b4d0f105c7

jarolrod reviewed (commented):
tACK 8621fbf672157df5475fc5cbaa6524b4d0f105c7

Generated some blocks on regtest and checked the output of `listunspent {args} ... include_immature_coinbase=true`, to check that various immature coinbases were infact displayed

jarolrod commented on test/functional/wallet_balance.py:
```diff
@@ -74,8 +74,18 @@ def run_test(self):
         self.generate(self.nodes[0], 1)
         self.generate(self.nodes[1], 1)
+
+        # Verify listunspent returning immature coinbase if 'include_immature_coinbase' is set
```
nit
```suggestion
        # Verify listunspent returns immature coinbase if 'include_immature_coinbase' is set
```

jarolrod commented on test/functional/wallet_balance.py:
```diff
@@ -74,8 +74,18 @@ def run_test(self):
+
         self.generatetoaddress(self.nodes[1], COINBASE_MATURITY + 1, ADDRESS_WATCHONLY)
 
+        # Verify listunspent returning all immature coinbases if 'include_immature_coinbase' is set
```
```suggestion
        # Verify listunspent returns all immature coinbases if 'include_immature_coinbase' is set
```

furszy commented:
thanks for the feedback.
Tackled @jarolrod nits and add release-notes.

jarolrod commented on doc/release-notes.md:
```diff
@@ -99,6 +99,10 @@ Wallet
   in wallet" (-4), if the label is not in the address book. (#25122)
 
+- RPC `listunspent` now has a new argument `include_immature_coinbase`
+  to include coinbase UTXOs that don't met the minimum spendability 
```
nit 
```suggestion
  to include coinbase UTXOs that don't meet the minimum spendability 
```

furszy replied:
should hire you as translator, pushed 👍🏼.

jarolrod reviewed (commented):
ACK ee50c00a1b3f09234a43f0fca140fcdd5f875ebd

Only change is small comment change and addition of release notes since my last review.

luke-jr commented on src/wallet/rpc/coins.cpp:
```diff
@@ -521,6 +521,9 @@ RPCHelpMan listunspent()
                         },
                         "query_options"},
+                    {
+                        "include_immature_coinbase", RPCArg::Type::BOOL, RPCArg::Default{false}, "Include coinbase UTXOs that haven't met the minimum spendability depth requirement\n"
```
Please don't add booleans as positional arguments. This should probably be part of `query_options` above.

furszy replied:
sure thanks, that is better.

kouloumos reviewed (commented):
Concept ACK,

Although this flag is already part of the `*receivedby*` RPCs since https://github.com/bitcoin/bitcoin/pull/14707, the logic there is different. Therefore this PR changes `AvailableCoins` (mainly [used by `CreateTransactionInternal` to get the available coins](https://github.com/bitcoin/bitcoin/blob/fc44d1796e4df5824423d7d13de3082fe204db7d/src/wallet/spend.cpp#L898-L905) prior to coin selection) in order to allow immature coinbase txns to make it to the available coins list. To accomplish that, it introduces the `include_immature_coinbase` option as part of coin control. 

Is the inclusion of immature coinbase txns in scope for Coin Control features? Is there a create-tx scenario that we will want to choose those txns?

achow101 commented:
Reopened by request

furszy committed:
wallet: group AvailableCoins filtering parameters in a single struct

Plus clean callers that use the params default values

furszy commented:
> Is the inclusion of immature coinbase txns in scope for Coin Control features? Is there a create-tx scenario that we will want to choose those txns?

Good point @kouloumos (sorry for the delay). It shouldn't be there. Just moved it to be another `AvailableCoins`  argument.

Plus, to make `AvailableCoins` more "friendly" to use with so many arguments, have grouped all the filtering parameters inside a struct 9f57166.

vladimirfomene reviewed (commented):
Review ACK. Just Nits

furszy commented on src/wallet/rpc/coins.cpp:
```diff
@@ -515,6 +515,7 @@ RPCHelpMan listunspent()
                             {"maximumAmount", RPCArg::Type::AMOUNT, RPCArg::DefaultHint{"unlimited"}, "Maximum value of each UTXO in " + CURRENCY_UNIT + ""},
                             {"maximumCount", RPCArg::Type::NUM, RPCArg::DefaultHint{"unlimited"}, "Maximum number of UTXOs"},
                             {"minimumSumAmount", RPCArg::Type::AMOUNT, RPCArg::DefaultHint{"unlimited"}, "Minimum sum value of all UTXOs in " + CURRENCY_UNIT + ""},
+                            {"include_immature_coinbase", RPCArg::Type::BOOL, RPCArg::Default{false}, "Include immature coinbase transactions"}
```
yep, same concept. If the tx is immature, all its outputs are immature. But let me push this.

kouloumos replied:
It's not that a big of an issue, but "transactions" is how it is used on all other RPCs that have the same flag, therefore this change creates a small inconsistency.

furszy replied:
yeah @kouloumos, I thought about it as well. Ended up changing it because in this RPC command `listunspent`, we only care about returning outputs. So, kinda make sense to use the UTXO term here.

Either way is fine for me.

kouloumos replied:
Agree, I didn't think it through.

aureleoules commented on src/wallet/spend.cpp:
```diff
@@ -301,19 +296,13 @@ CoinsResult AvailableCoins(const CWallet& wallet,
 
 CoinsResult AvailableCoinsListUnspent(const CWallet& wallet, const CCoinControl* coinControl, const CAmount& nMinimumAmount, const CAmount& nMaximumAmount, const CAmount& nMinimumSumAmount, const uint64_t nMaximumCount, bool include_immature_coinbase)
```
427548d8df0b2786ee360f443bb07adc264773bc
maybe you could wrap these args as well?

furszy replied:
I didn't do it because of the `only_spendable` arg that must be true for this method (otherwise shouldn't be called "AvailableCoins**ListUnspent**").
But.. could also pass the struct and force set the `only_spendable` flag internally too.

kouloumos replied:
@furszy if you do decide to do that, maybe by using the struct here as well
https://github.com/bitcoin/bitcoin/blob/0bac04b7585017508425b21d402fd3c52e6a13d5/src/wallet/rpc/coins.cpp#L593-L596
it creates a good opportunity (as you'll now touch all the existing code that uses them) to [rename the variables to snake_case](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c)?

furszy replied:
Pushed both changes.

Squashed the `AvailableCoinsListUnspent` changes inside  56c7d7d and added an scripted-diff commit 1022c18 at the end for the snake_case rename to minimize the review burden.

aureleoules commented on src/wallet/spend.cpp:
```diff
@@ -895,13 +884,7 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
-                                              MAX_MONEY,    /*nMaximumAmount*/
-                                              MAX_MONEY,    /*nMinimumSumAmount*/
-                                              0);           /*nMaximumCount*/
+    auto available_coins = AvailableCoins(wallet, &coin_control, coin_selection_params.m_effective_feerate);
```
For reviewers: These hardcoded values can be removed as they are now the defaults of `AvailableCoinsParams`.

aureleoules reviewed (approved):
ACK 1406f1d0f1db5cd68c2459953ba913e845cb36e2 - LGTM

427548d8df0b2786ee360f443bb07adc264773bc: I verified that the calls with hard-coded values match the defaults of `AvailableCoinsParams`.

rajarshimaitra reviewed (approved):
Concept + tACK 427548d8df0b2786ee360f443bb07adc264773bc

Made some regtest blocks and manually created coinbase transactions and verified that they are reported in list unspent. This is in general useful for any wallet library that wants to sync with core using `listunspent`. 

ACK on grouping the `AvialbleCoins` args, looks much more cleaner..

furszy committed:
scripted-diff: wallet: rename AvailableCoinsParams members to snake_case

-BEGIN VERIFY SCRIPT-

sed -i 's/nMinimumAmount/min_amount/g' $(git grep -l nMinimumAmount)
sed -i 's/nMaximumAmount/max_amount/g' $(git grep -l nMaximumAmount)
sed -i 's/nMinimumSumAmount/min_sum_amount/g' $(git grep -l nMinimumSumAmount)
sed -i 's/nMaximumCount/max_count/g' $(git grep -l nMaximumCount)

-END VERIFY SCRIPT-

furszy commented:
updated per feedback

rajarshimaitra reviewed (approved):
ReACK 14242469930701462e39e34525fbd1a32a16daa7

aureleoules reviewed (approved):
reACK 14242469930701462e39e34525fbd1a32a16daa7

kouloumos commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1385,7 +1385,9 @@ RPCHelpMan sendall()
                 }
             } else {
-                for (const COutput& output : AvailableCoins(*pwallet, &coin_control, fee_rate, /*nMinimumAmount=*/0).All()) {
+                AvailableCoinsParams coins_params;
```
nit: Regarding consistency for variable names, I think the `filter_coins` name that you use for the same variable in `wallet/rpc/coins.cpp` is better, or maybe a mixture of your current names: `coins_filter_params`.

Talking about naming, although `AvailableCoinsParams` is only used in the `AvailableCoins` function (hence the naming I guess), I think the name could do a better job describing the content. May I suggest `CoinFilter(ing)Params`, which also follows the naming scheme used in `CoinSelectionParams`.
https://github.com/bitcoin/bitcoin/blob/3f1f5f6f1ec49d0fb2acfddec4021b3582ba0553/src/wallet/coinselection.h#L116

furszy replied:
thought about it too. I tried to use a distinctive name for this struct because we do have some "coins filtering" args inside `CoinControl` as well.

But yeah, probably the best is to use the `CoinFilterParams` naming for this new struct and, in a future PR, instead of passing `CoinControl` to `AvailableCoins`, only pass `CoinFilterParams` (which will need to include what is inside `CoinControl`).

kouloumos replied:
I have to admit that I haven't look deep into the usage of CoinControl yet, but I I don't see such an issue on the distinction that `CoinFilterParams` are all the filters that are not part of `CoinControl`. Cause CoinControl is also a user-facing concept.

furszy replied:
What I meant is that, ideally, we should only provide `CoinFilterParams` to `AvailableCoins` and not two structs that contain filtering args. But that can be done later (I'm removing the preset inputs skip from `AvailableCoins` inside #25685).

kouloumos reviewed (commented):
ACK 14242469930701462e39e34525fbd1a32a16daa7, with a small nit about naming.

furszy commented:
Updated per @kouloumos feedback.

tiny [diff](https://github.com/bitcoin/bitcoin/compare/14242469930701462e39e34525fbd1a32a16daa7..3319eb9d3238f6fc02d51694ca5f215c58b7fc0b), only renamed `AvailableCoinsParams` to `CoinFilterParams`.

kouloumos commented:
reACK 3319eb9d3238f6fc02d51694ca5f215c58b7fc0b

aureleoules reviewed (approved):
reACK 3319eb9d3238f6fc02d51694ca5f215c58b7fc0b

danielabrozzoni commented on test/functional/wallet_balance.py:
```diff
@@ -77,8 +77,18 @@ def run_test(self):
+        # Verify listunspent returns all immature coinbases if 'include_immature_coinbase' is set
+        # For now, only the legacy wallet will see the coinbases going to the imported 'ADDRESS_WATCHONLY'
+        assert_equal(len(self.nodes[0].listunspent(query_options={'include_immature_coinbase': False})), 1 if self.options.descriptors else 2)
+        assert_equal(len(self.nodes[0].listunspent(query_options={'include_immature_coinbase': True})), 1 if self.options.descriptors else COINBASE_MATURITY + 2)
```
tiny tiny nit: the `COINBASE_MATURITY + 2` left me wondering for a sec, this might be easier to read if we had a variable storing how many blocks were generated:
```
n_blocks_generate = COINBASE_MATURITY + 2  # feel free to use a better name :)
self.generatetoaddress(self.nodes[1],  n_blocks_generate, ADDRESS_WATCHONLY)
...
assert_equal(len(self.nodes[0].listunspent(query_options={'include_immature_coinbase': True})), 1 if self.options.descriptors else n_blocks_generate + 1)
```

furszy replied:
k, as it is a tiny tiny nit, will only do it if have to retouch the code so we don't loose the ACKs.

danielabrozzoni commented on src/wallet/spend.h:
```diff
@@ -55,23 +55,34 @@ struct CoinsResult {
 
 /**
- * Wrapper function for AvailableCoins which skips the `feerate` parameter. Use this function
+ * Wrapper function for AvailableCoins which skips the `feerate` and `CoinFilterParams::only_spendable` parameters. Use this function
```
Sorry, I'm not sure I follow here 😅 

Why are you saying that this function "skips" `CoinFilterParams::only_spendable`? To me, it seems that it's skipping only the `feerate` parameter (i.e., `AvailableCoins` takes a `feerate` param, and `AvailableCoinsListUnspent` doesn't)

furszy replied:
Internally, `AvailableCoinsListUnspent` sets `only_spendable` to false (this isn't a new behavior introduced on this PR, `AvailableCoinsListUnspent` callers expect to receive all the coins). Check one of my previous comments https://github.com/bitcoin/bitcoin/pull/25730#discussion_r994815499

So, what the comment basically says is "do not customize `CoinFilterParams::only_spendable`, it will not be taken into account".

danielabrozzoni replied:
Ah, now it makes sense, thanks!

danielabrozzoni reviewed (commented):
Concept ACK, I have a couple of questions (beginner here, sorry :)) and I still haven't tested the code, but overalls it looks good to me!

furszy commented on src/wallet/spend.cpp:
```diff
@@ -299,21 +294,16 @@ CoinsResult AvailableCoins(const CWallet& wallet,
-            /*nMinimumSumAmount=*/ MAX_MONEY,
-            /*nMaximumCount=*/ 0
-    ).total_amount;
+    return AvailableCoins(wallet, coinControl).total_amount;
```
because `feerate` default value is `std::nullopt` (see `AvailableCoins` declaration). So, no need to explicitly set it.

danielabrozzoni reviewed (approved):
tACK 3319eb9d3238f6fc02d51694ca5f215c58b7fc0b - the code looks good to me (but my C++ experience is limited), I tested locally and it works as expected. As I pointed out in #25728, this makes it way easier to track immature coins - at the moment in BDK we have to use `listunspent` combined with `listtransactions` to make sure to track correctly immature utxos.

Thanks for taking care of this! :)

theStack commented:
Concept ACK

furszy commented:
Rebased, conflicts solved.
Ready to go.

aureleoules reviewed (approved):
reACK fa84df1f033a5d1a8342ea941eca0b5ef73d78e7

kouloumos reviewed (commented):
reACK fa84df1f033a5d1a8342ea941eca0b5ef73d78e7

danielabrozzoni commented:
reACK fa84df1f033a5d1a8342ea941eca0b5ef73d78e7

theStack reviewed (approved):
Code-review ACK fa84df1f033a5d1a8342ea941eca0b5ef73d78e7
